File: easy-scala/src/main/scala/com/xiaodong/spark/easy/scala/RandomExample.scala
package com.xiaodong.spark.easy.scala

import scala.math.random

/**
  * Created by lixiaodong on 16/9/28.
  */
object RandomExample {
    def main(args: Array[String]): Unit = {
        println(piPositive(200000000))
        println(pi(200000000))
    }

    def pi(slices: Int): Double = {
        val n = math.min(slices, Int.MaxValue)
        var count:Double = 0
        for (i <- 0 until n) {
            val x = random * 2 - 1
            val y = random * 2 - 1
            if (x*x + y*y < 1)
                count += 1
        }
        4 * count/n
    }

    /**
      * 坐标x,y都是大于0的
      * @param slices
      * @return
      */
    def piPositive(slices: Int):Double = {
        val n = math.min(slices, Int.MaxValue)
        var count:Double = 0
        for (i <- 0 until n) {
            val x = random
            val y = random
            if (x*x + y*y < 1)
                count += 1
        }
        4 * count/n
    }
}

File: easy-scala/src/main/java/com/xiaodong/spark/easy/java/RandomExample.java
package com.xiaodong.spark.easy.java;

import java.util.Random;

/**
 * Created by lixiaodong on 16/9/28.
 */
public class RandomExample {
    public static void main(String[] args) {
        System.out.println(pi(20000000));
        System.out.println(piPositive(20000000));
    }

    private static double piPositive(int slices) {
        double count = 0;
        Random random = new Random();
        for (int i=0; i<slices; i++) {
            double x = random.nextDouble();
            double y = random.nextDouble();
            if (x*x + y*y < 1) {
                count ++;
            }
        }
        return 4.0 * count/slices;
    }

    private static double pi(int slices) {
        double count = 0;
        Random random = new Random();
        for (int i=0; i<slices; i++) {
            double x = random.nextDouble() * 2.0 - 1;
            double y = random.nextDouble() * 2.0 - 1;
            if (x*x + y*y < 1) {
                count ++;
            }
        }
        return 4.0 * count/slices;
    }
}
